File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Middleware/ConsoleOutputMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Bot.Solutions.Testing
{
    using System;
    using System.Collections.Generic;
    using System.Threading;
    using System.Threading.Tasks;
    using Microsoft.Bot.Builder;
    using Microsoft.Bot.Schema;

    public class ConsoleOutputMiddleware : IMiddleware
    {
        public async Task OnTurnAsync(ITurnContext context, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            this.LogActivity(string.Empty, context.Activity);
            context.OnSendActivities(this.OnSendActivities);

            await next(cancellationToken).ConfigureAwait(false);
        }

        private static string GetTextOrSpeak(IMessageActivity messageActivity)
        {
            return string.IsNullOrWhiteSpace(messageActivity.Text) ? messageActivity.Speak : messageActivity.Text;
        }

        private async Task<ResourceResponse[]> OnSendActivities(ITurnContext context, List<Activity> activities, Func<Task<ResourceResponse[]>> next)
        {
            foreach (var response in activities)
            {
                this.LogActivity(string.Empty, response);
            }

            return await next();
        }

        private void LogActivity(string prefix, Activity contextActivity)
        {
            Console.WriteLine(string.Empty);
            if (contextActivity.Type == ActivityTypes.Message)
            {
                var messageActivity = contextActivity.AsMessageActivity();
                Console.WriteLine($"{prefix} [{DateTime.Now:ss.fff}] {GetTextOrSpeak(messageActivity)}");
            }
            else if (contextActivity.Type == ActivityTypes.Event)
            {
                var eventActivity = contextActivity.AsEventActivity();
                Console.WriteLine($"{prefix} Event: [{DateTime.Now:ss.fff}{DateTime.Now.Millisecond}] {eventActivity.Name}");
            }
            else
            {
                Console.WriteLine($"{prefix} {contextActivity.Type}: [{DateTime.Now:ss.fff}{DateTime.Now.Millisecond}]");
            }
        }
    }
}

File: sdk/typescript/libraries/botbuilder-solutions/src/middleware/consoleOutputMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Activity, ActivityTypes, Middleware, TurnContext  } from 'botbuilder';
import { IEventActivity, IMessageActivity, ResourceResponse } from 'botframework-schema';

export class ConsoleOutputMiddleware implements Middleware {
    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        this.logActivity('', context.activity);
        context.onSendActivities(this.onSendActivities.bind(this));

        await next();
    }

    private static getTextOrSpeak(messageActivity: IMessageActivity): string {
        if (messageActivity.text === undefined || messageActivity.text === '') {
            return messageActivity.speak || '';
        }

        return messageActivity.text;
    }

    private async onSendActivities(
        context: TurnContext,
        activities: Partial<Activity>[],
        next: () => Promise<ResourceResponse[]>
    ): Promise<ResourceResponse[]> {
        activities.forEach((response: Partial<Activity>): void => {
            this.logActivity('', response);
        });

        return next();
    }

    private logActivity(prefix: string, contextActivity: Partial<Activity>): void {
        this.log('');
        if (contextActivity.type === ActivityTypes.Message) {
            const messageActivity: IMessageActivity = contextActivity as IMessageActivity;
            this.log(`${ prefix } [${ Date.now() }] ${ ConsoleOutputMiddleware.getTextOrSpeak(messageActivity) }`);
        } else if (contextActivity.type === ActivityTypes.Event) {
            const eventActivity: IEventActivity = contextActivity as IEventActivity;
            this.log(`${ prefix } [${ Date.now() }] ${ eventActivity.name }`);
        } else {
            this.log(`${ prefix } ${ contextActivity.type }: [${ Date.now() }]`);
        }
    }

    private log(message: string): void {
        console.log(message);
    }
}
